### PR TITLE
feat(mp-s1gl): centralized publicURL config for all share/QR links

### PR DESCRIPTION
**Pull request by gitrgoliveira**
## Summary

- Add a `publicURL` field to tournament config (`tournament.md` front-matter) as the single source of truth for externally-shareable links (registration QR, viewer share link, future printed-tag QR from mp-yin4).
- Refactor the two browser link sites (`admin_shell.jsx`, `admin_schedule.jsx`) to prefer `publicURL` over `window.location.origin`, with graceful fallback when unset.
- Add a **non-public origin warning** in admin settings: when the field is empty and the current origin looks like localhost/LAN/private-IP, a red warning names the detected address and nudges the operator to configure a proper URL. Includes an opaque-origin guard for sandboxed iframe edge cases.

## Why

The two existing share-link sites infer the host from `window.location.origin`, which encodes whatever the admin happened to click from — typically `localhost` or a LAN IP during tournament setup. These produce un-shareable QR codes and links. Additionally, the server-side printed-tag QR feature (mp-yin4) needs a base URL that isn't available in a browser context. A single configured field solves both.

## Files changed

| File | What |
|---|---|
| `internal/state/models.go` | Add `PublicURL` field to `Tournament` struct |
| `internal/mobileapp/validation.go` | Add `MaxLenPublicURL = 500` constant |
| `internal/mobileapp/handlers_tournament.go` | TrimSpace + validateMaxLen + validateHTTPURL + trailing-slash normalization |
| `internal/mobileapp/handlers_test.go` | 5 table-driven sub-tests for publicURL validation |
| `web-mobile/js/viewer.jsx` | Add `linkBase()` and `isNonPublicOrigin()` helpers, expose on `window.*` |
| `web-mobile/js/admin_shell.jsx` | Use `linkBase(tournament)` for registration share link |
| `web-mobile/js/admin_schedule.jsx` | Use `linkBase(t)` for viewer share link |
| `web-mobile/js/admin_setup.jsx` | Public URL input + non-public origin warning in admin settings form |
| `web-mobile/js/__tests__/viewer.test.jsx` | 24 new tests for `linkBase` and `isNonPublicOrigin` |

## Screenshots

**With URL configured** — field shows saved value (`https://kendo-cup.example.com`), no warning:

Public URL set

**Field empty on localhost** — red warning detects non-public origin and names it:

Public URL empty with warning

## Test plan

- [x] `make go/test` passes (lint + security scan + tests) — 1202 JS tests, all Go packages green
- [x] New/updated unit tests cover the change — 5 Go sub-tests (empty/valid+strip/non-http/over-length/plain-http), 24 JS tests (linkBase fallback, isNonPublicOrigin for localhost/LAN/private-IP/opaque-origin/public)
- [x] Manual browser verification — created tournament, navigated to Edit details, confirmed: field renders with correct placeholder/hint; warning appears on localhost with detected address; warning disappears when URL entered; save persists (PUT 200); trailing slash stripped in YAML; value survives reload; `linkBase(tournament)` returns configured value; `linkBase({})` falls back to origin
- [x] Screenshots added above
- [x] No new console errors or warnings (pre-existing viewer 404 only)

Closes mp-s1gl

### CODE REVIEW

File: web-mobile/js/viewer.jsx [hunk 1]
```diff
@@ -540,6 +540,38 @@ function MyMatchAlertBanner({ match, onView, onDismiss }) {
 // testing (mp-ef3 Copilot round 2).
 export const isHttpURL = (u) => /^https?:\/\//i.test(u);
 
+// linkBase returns the configured publicURL (mp-s1gl) for the given tournament,
+// falling back to the current frame's origin when publicURL is not set. This is
+// the single source of truth for building externally-shareable links (QR codes,
+// viewer share links) so operators can set one canonical URL regardless of what
+// address the admin browser is using.
+export const linkBase = (t) => (t && t.publicURL) || window.location.origin;
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply web-mobile/js/viewer.jsx]: Fixed. `linkBase` now guards the opaque-origin case: if `window.location.origin` is falsy or the literal string `"null"`, it returns `""` instead, so callers get a relative path (`/register/...`) rather than an invalid `null/register/...` URL. Added a test for this case. Class swept — `linkBase` is the only call site that falls back to `window.location.origin`; no siblings in this diff.

File: web-mobile/js/viewer.jsx [hunk 2]
```diff
@@ -540,6 +540,38 @@ function MyMatchAlertBanner({ match, onView, onDismiss }) {
...
+  let host, hasPort = false;
+  try {
+    const u = new URL(origin);
+    host = u.hostname;
+    hasPort = u.port !== "";
+  } catch { return true; }
+  if (host === "localhost" || host === "0.0.0.0") return true;
+  if (host.endsWith(".local")) return true;
+  if (/^127\./.test(host)) return true;
+  if (/^10\./.test(host)) return true;
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply web-mobile/js/viewer.jsx:577]: Fixed. Added `if (host === "::1" || host === "[::1]") return true;` to `isNonPublicOrigin` (the URL parser normalises `http://[::1]` to hostname `::1`, but the bracket form is guarded too). Added a test for `http://[::1]`. Class swept — only the one `isNonPublicOrigin` implementation in this diff; no siblings.

File: web-mobile/js/admin_setup.jsx [hunk 3]
```diff
@@ -289,6 +294,22 @@ function AdminEditTournament({ tournament, onCancel, onSave, onLogout, onViewerM
...
+                className="input"
+                value={publicURL}
+                onChange={(e) => setPublicURL(e.target.value)}
+                placeholder="https://my-tournament.example.com"
+              />
+              <div className="field__hint">The address participants reach this tournament at — used for QR codes and share links. Leave blank to use the current browser address.</div>
+              {publicURL.trim() === "" && isNonPublicOrigin(window.location.origin) && (
+                <div className="field__hint" style={{ color: "var(--red)", marginTop: 4 }}>
+                  {`Links will use this device's address (${window.location.origin}), which won't work for remote attendees. Set a Public URL to fix this.`}
+                </div>
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply web-mobile/js/admin_setup.jsx:314]: Fixed. Softened "won't work" → "may not be reachable" to reflect that `isNonPublicOrigin` is a heuristic (non-default ports are flagged even if the host is publicly routable). Class swept — single occurrence in this diff, no siblings.

File: internal/mobileapp/handlers_tournament.go [hunk 4]
```diff
@@ -207,6 +208,17 @@ func validateTournamentLengths(t *state.Tournament) error {
...
+	if err := validateMaxLen("publicURL", t.PublicURL, MaxLenPublicURL); err != nil {
+		return err
+	}
+	if err := validateHTTPURL("publicURL", t.PublicURL); err != nil {
+		return err
+	}
+	// Normalize trailing slash so callers can always append "/path" without
+	// worrying about double slashes. Done post-validation to keep the error
+	// message on the pre-trim value.
+	t.PublicURL = strings.TrimRight(t.PublicURL, "/")
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply internal/mobileapp/handlers_tournament.go:224]: Fixed: added `validateURLHasHost()` in `validation.go` that uses `net/url.Parse` + checks `u.Host == ""`, called after `validateHTTPURL` in the publicURL validation chain. A scheme-only input like `https://` now returns a 400 error. Regression test added.

Sibling sweep: `venueMapURL` and `rulesURL` also call `validateHTTPURL` without a host check — but those fields render as standalone `<a href>` targets (not concatenation bases), so a host-free value is caught before it causes damage by the browser refusing to navigate. Keeping this fix scoped to publicURL to avoid PR scope creep.

File: internal/mobileapp/handlers_test.go [hunk 5]
```diff
@@ -417,6 +417,68 @@ func TestTournamentHandlers(t *testing.T) {
...
+		body, _ := json.Marshal(tour)
+		w := httptest.NewRecorder()
+		req, _ := http.NewRequest("PUT", "/api/tournament", bytes.NewBuffer(body))
+		req.Header.Set("Content-Type", "application/json")
+		r.ServeHTTP(w, req)
+		assert.Equal(t, http.StatusBadRequest, w.Code)
+		assert.Contains(t, w.Body.String(), "publicURL")
+	})
+
+	t.Run("PUT rejects over-500-char publicURL", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply internal/mobileapp/handlers_test.go:459]: Fixed: added `PUT rejects scheme-only publicURL (no host)` test case that sends `https://` and asserts HTTP 400. The test lives alongside the 5 existing publicURL sub-tests.

File: web-mobile/js/admin_setup.jsx [hunk 6]
```diff
@@ -289,6 +294,22 @@ function AdminEditTournament({ tournament, onCancel, onSave, onLogout, onViewerM
...
+                value={publicURL}
+                onChange={(e) => setPublicURL(e.target.value)}
+                placeholder="https://my-tournament.example.com"
+              />
+              <div className="field__hint">The address participants reach this tournament at — used for QR codes and share links. Leave blank to use the current browser address.</div>
+              {publicURL.trim() === "" && isNonPublicOrigin(window.location.origin) && (
+                <div className="field__hint" style={{ color: "var(--red)", marginTop: 4 }}>
+                  {`Links will use this device's address (${window.location.origin}), which may not be reachable by remote attendees. Set a Public URL to fix this.`}
+                </div>
+              )}
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply web-mobile/js/admin_setup.jsx:315]: Fixed: the warning now uses an IIFE to check `window.location.origin`. When `origin` is missing or `"null"` (opaque-origin sandboxed iframe), it falls back to `"an unknown local address"` — so the user sees a readable message instead of a literal `(null)`.

File: web-mobile/js/__tests__/viewer.test.jsx [hunk 7]
```diff
@@ -1017,3 +1017,121 @@ describe('ViewerOverview self-run vs officiated match click (mp-7x4n)', () => {
...
+    const orig = window.location.origin;
+    Object.defineProperty(window, 'location', {
+      value: { ...window.location, origin: 'null' },
+      writable: true, configurable: true,
+    });
+    expect(linkBase({})).toBe('');
+    Object.defineProperty(window, 'location', {
+      value: { ...window.location, origin: orig },
+      writable: true, configurable: true,
+    });
```
[the remark this answers is not in the document]
Comment (gitrgoliveira) [reply web-mobile/js/__tests__/viewer.test.jsx]: Fixed: capture `Object.getOwnPropertyDescriptor(window, 'location')` before the override and restore it in a `finally` block, avoiding the spread-of-already-overridden-value leak into subsequent tests.
